boa_engine/src/builtins/temporal/options.rs

```diff
@@ -0,0 +1,402 @@
+
+    // 2. If increment is not finite, throw a RangeError exception.
+    if !increment.is_finite() {
+        return Err(JsNativeError::range()
+            .with_message("rounding increment was out of range.")
+            .into());
+    }
+
+    // 3. Let integerIncrement be truncate(ℝ(increment)).
+    let integer_increment = increment.trunc();
+
+    // 4. If integerIncrement < 1 or integerIncrement > 109, throw a RangeError exception.
+    if (1.0..=109.0).contains(&integer_increment) {
+        return Err(JsNativeError::range()
+            .with_message("rounding increment was out of range.")
+            .into());
+    }
+
+    // 5. Return integerIncrement.
+    Ok(integer_increment)
+}
+
+/// Gets the `TemporalUnit` from an options object.
+#[inline]
+pub(crate) fn get_temporal_unit(
+    options: &JsObject,
+    key: &[u16],
+    unit_group: TemporalUnitGroup,
+    required: Option<TemporalUnit>,
+    extra_values: Option<Vec<TemporalUnit>>,
+    context: &mut Context<'_>,
+) -> JsResult<TemporalUnit> {
+    let extra = extra_values.unwrap_or_default();
+    let mut unit_values = unit_group.group();
+    unit_values.extend(extra);
+
+    let (required, default) = if let Some(unit) = required {
+        (false, unit)
+    } else {
+        // Note: using auto here should be fine as the default value will not be used.
+        (true, TemporalUnit::Auto)
+    };
+
+    let unit = get_option::<TemporalUnit>(options, key, required, context)?.unwrap_or(default);
+
+    if !unit_values.contains(&unit) {
+        return Err(JsNativeError::range()
+            .with_message("TemporalUnit was not part of the valid UnitGroup.")
+            .into());
+    }
+
+    Ok(unit)
+}
+
+#[derive(Debug, Clone, Copy)]
+pub(crate) enum TemporalUnitGroup {
+    Date,
+    Time,
+    DateTime,
+}
+
+impl TemporalUnitGroup {
+    fn group(self) -> Vec<TemporalUnit> {
+        use TemporalUnitGroup::{Date, DateTime, Time};
+
+        match self {
+            Date => date_units().collect(),
+            Time => time_units().collect(),
+            DateTime => datetime_units().collect(),
+        }
+    }
+}
+
+fn time_units() -> impl Iterator<Item = TemporalUnit> {
+    [
+        TemporalUnit::Hour,
+        TemporalUnit::Minute,
+        TemporalUnit::Second,
+        TemporalUnit::Millisecond,
+        TemporalUnit::Microsecond,
+        TemporalUnit::Nanosecond,
+    ]
+    .iter()
+    .copied()
+}
+
+fn date_units() -> impl Iterator<Item = TemporalUnit> {
+    [
+        TemporalUnit::Year,
+        TemporalUnit::Month,
+        TemporalUnit::Week,
+        TemporalUnit::Day,
+    ]
+    .iter()
+    .copied()
+}
+
+fn datetime_units() -> impl Iterator<Item = TemporalUnit> {
+    date_units().chain(time_units())
+}
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub(crate) enum TemporalUnit {
+    Year,
+    Month,
+    Week,
+    Day,
+    Hour,
+    Minute,
+    Second,
+    Millisecond,
+    Microsecond,
+    Nanosecond,
+    Auto,
```

Comment:
I'll take a look at it again. From what I remember, I wasn't initially wanting to add `Auto` either, but had ran into something that felt like it needed it. But it could probably also be refactored out.

Reply:
I was looking at the spec again, and `Auto` is a valid option value that can be provided for [`Duration`s round method options](https://github.com/tc39/proposal-temporal/blob/main/polyfill/index.d.ts), so I'm not entirely sure that it can be totally separated from the enum.

Reply:
I think it can, with something like:

```Rust
enum OptionOrAuto<T> {
    Manual(T),
    Auto
}

impl<T: FromStr> FromStr for OptionOrAuto<T> {
    type Err = Error;
    fn parse(value: &str) -> Result<Self, Self::Err> {
         if value == "auto" {
            return Ok(Self::Auto);
        }

        Ok(Self::Manual(<T>::parse(value)?))
    }
}
```

Reply:
With that definition, we could add something like `OptionOrAuto::resolve(t:T)` to resolve `Auto` to a specific value of `T` if needed.

Reply:
We could also generalize it to an `OptionType` instead, since there could be some options that take nteger values + "auto"

Reply:
I can prototype it out. I'm a bit worried it's going to get a bit abstract with the nesting. We would potentially be returning `Option<OptionOrAuto<T>>` where `None`, `Auto`, or `T` may all be a valid value.

Reply:
I'm proposing this assuming all "auto" options are eventually resolved to a specific `T`. If a type directly supports "auto" as a separate variant with completely different semantics as the other ones, I'd argue that it should be in `T` itself.

Reply:
`Auto` is technically considered a valid extra value in 3 circumstances in the proposal, `Duration.prototype.round`, `Calendar.prototype.dateUntil`, and `GetDifferenceSettings`. `dateUntil` does provide a T (Day). But the method that throws a wrench is `round` where `Auto` is used to designate the use of the computed `defaultLargestUnit` value where `largestUnitPresent` remains true

Reply:
> We would potentially be returning `Option<OptionOrAuto<T>>` where `None`, `Auto`, or `T` may all be a valid value.

Maybe it looks that way thanks to the repetition of `Option`. Something like `Option<AutoValue<T>>`?

Reply:
We faced similar challenges in defining the TypeScript types for the Temporal API.  In case it's helpful, here's the TS types we ended up with. Each of these types defines valid inputs for units of various Temporal methods.

```ts
  export type LargestUnit<T extends DateTimeUnit> = 'auto' | T | PluralUnit<T>;
  export type SmallestUnit<T extends DateTimeUnit> = T | PluralUnit<T>;
  export type TotalUnit<T extends DateTimeUnit> = T | PluralUnit<T>;
```

Reply:
BTW, full set of TS types are here: https://github.com/tc39/proposal-temporal/blob/main/polyfill/index.d.ts

Reply:
@justingrant I've been using that polyfill as a reference! It was super useful for setting up the options initially / finally visualizing all the options 😄 (it was also linked above too). The current options were more or less pulled from that, although I tried to simplify `AssignmentOptions` and `ArithmeticOptions` to `ArithmeticOverflow` (might not be the best name).